### PR TITLE
Added genereic REST API action

### PR DESCRIPTION
**Pull request by deinhofer**
Added genereic REST API action (as of #76 ), which currently supports:

- Setting of a REST URL
- Setting of body (data) value

e..g.
image

## not yet supported

- custom method
- custom content-type

### CODE REVIEW

File: src/js/service/restService.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        if(method!='GET') {
+            body=action.body;
+        }
+
+        console.log(`url: ${action.restUrl}, body: ${body}, method: ${method}, contenttype: ${contentType}`);
+        const response = await fetch(action.restUrl, {
+                method: method,
+                //mode: "no-cors",
+                headers: {
+                    'Content-Type': contentType
```
Comment (klues) [inline src/js/service/restService.js:30]: `'Content-Type': action.contentType || GridActionREST.defaults.contentType`

File: src/js/service/restService.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        } else {
+            //console.log(`response body: ${response.json().then()}`);
+            log.debug(`REST call ok, url: ${action.restUrl}, body ${action.body}`)
+        }
+    }catch(error) {
+        console.error(error);
+    }
+};
+
+restService.getRestURL = function (userUri) {
```
Comment (klues) [inline src/js/service/restService.js:46]: I think there should be something like a generic `util.getRestURL(userUrl, options)` which then is used here, and at `openHABService.getRestURL()` and `areService.getRestURL()`

File: src/vue-components/modals/editActionsSub/editRESTAction.vue [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+}
+</script>
+
+<style scoped>
+
+.normal-text {
+    font-weight: normal;
+}
+
+.custom-radio {
```
Comment (klues) [inline src/vue-components/modals/editActionsSub/editRESTAction.vue:73]: please check which of these custom CSS rules are actually used by your component

**Review by klues — CHANGES_REQUESTED** [review]
Please also add documentation for the new action. I think at some point we should split [05_actions.md](https://github.com/asterics/AsTeRICS-Grid/blob/master/docs/documentation_user/05_actions.md) into multiple files.

File: src/js/service/restService.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import { util } from '../util/util';
+import {GridActionREST} from "../model/GridActionREST";
+
+let restService = {};
+
+restService.doAction = async function (action) {
+    try {
+        //ensure to have default values also in case the user deletes the field contents
+        let method=action.method;
+        if(!method) {
```
[the remark this answers is not in the document]
Comment (deinhofer) [reply src/js/service/restService.js:10]: what exactly should be changed in the formatting?
```ctrl+alt+l``` does not change anything.

File: src/vue-components/modals/editActionsSub/editRESTAction.vue (hunk 3 above)
Comment (deinhofer) [reply src/vue-components/modals/editActionsSub/editRESTAction.vue:73]: ok, removed. 
Actually I think the CSS rules should also be encapsulated into a super vue component, as now it's copied for every component which makes changes very hard.

File: src/js/service/restService.js (hunk 2 above)
Comment (deinhofer) [reply src/js/service/restService.js:46]: will remove this method as it's not suitable for a generic REST method as there are too many variables.

**Comment by klues** [comment]
I've reviewed everything and did these changes:
* renamed "REST action" to "HTTP action" because I think it's clearer and more generic
* removed some translated strings, since they are the same for all languages, e.g. "HTTP Content-Type", see https://developer.mozilla.org/zh-CN/docs/Web/HTTP/Headers/Content-Type
* using a `<select>` element for HTTP method
* using a `<textarea>` element for HTTP body (for being able to define longer payloads
* removed some superfluous HTML in `editHttpAction.vue`
* added documentation, see https://github.com/asterics/AsTeRICS-Grid/blob/latest/docs/documentation_user/05_actions.md#http-action

Please test everything again at https://grid.asterics.eu/latest/ - I think then we could release it to the main version 🥳

**Comment by klues** [comment]
I also did tests:
* [httpbin](https://httpbin.org/#/Auth/get_basic_auth__user___passwd_) for testing authorization. In general it worked, however not for users/passwords including umlauts - but I think that's an error of httpbin.
* ARE REST API, sending a model using `PUT` and `text/xml` - worked!

**Comment by deinhofer** [comment]
> I've reviewed everything and did these changes:
> 
>     * renamed "REST action" to "HTTP action" because I think it's clearer and more generic
> 
>     * removed some translated strings, since they are the same for all languages, e.g. "HTTP Content-Type", see https://developer.mozilla.org/zh-CN/docs/Web/HTTP/Headers/Content-Type
> 
>     * using a `<select>` element for HTTP method
> 
>     * using a `<textarea>` element for HTTP body (for being able to define longer payloads
> 
>     * removed some superfluous HTML in `editHttpAction.vue`
> 
>     * added documentation, see https://github.com/asterics/AsTeRICS-Grid/blob/latest/docs/documentation_user/05_actions.md#http-action
> 
> 
> Please test everything again at https://grid.asterics.eu/latest/ - I think then we could release it to the main version 🥳

Hi,

thanks for reviewing and finalizing!! It looks great!!
There is only one thing regarding terminology:

'HTTP Authorization User' and 'HTTP Authorization Password' should 
'HTTP **Authentication** User' and 'HTTP **Authentication** Password' as it's about authenticating users' credentials, e.g. is it really user A?

Authorization is more about granting different types of operations e.g. read or write.

But besides of that, I think it's read to merge!! :+1:

**Comment by deinhofer** [comment]
> I also did tests:
> 
>     * [httpbin](https://httpbin.org/#/Auth/get_basic_auth__user___passwd_) for testing authorization. In general it worked, however not for users/passwords including umlauts - but I think that's an error of httpbin.
> 
>     * ARE REST API, sending a model using `PUT` and `text/xml` - worked!

Do you think we should also add some REST API examples?
E.g. for the shelly plug we could describe the following examples:
```
http://192.168.33.1/relay/0?turn=toggle
http://192.168.33.1/relay/0?turn=off
http://192.168.33.1/relay/0?turn=on
```

If we documented some cases like this, we would probably don't need to integrate each API with another HTTP action.

**Comment by klues** [comment]
Thanks for testing! I've changed `Authorization` to `Authentication` and released it again to latest. I think we now can release this version to main - I'll do it after the last tests for the word forms feature (#254) are finished.

Regarding the templates - good idea - maybe I'll add it in future releases (we should collect some sensible examples before).

Thanks for your contribution! 👍

**Comment by deinhofer** [comment]
> Thanks for testing! I've changed `Authorization` to `Authentication` and released it again to latest. I think we now can release this version to main - I'll do it after the last tests for the word forms feature (#254) are finished.
> 
> Regarding the templates - good idea - maybe I'll add it in future releases (we should collect some sensible examples before).
> 
> Thanks for your contribution! 👍

Great!! Thanks!!

Yes, let's finally merge!

**Pull request merged**